Support for IPv6 subnets

First of all, thanks for making this. This is a really interesting project.
I was wondering how hard it would be to support IPv6-only subnets. By the looks of it, only https://github.com/skx/simple-vpn/blob/master/cmd_server.go#L170 assumes IPv4 and that's because we treat it like a string instead of like an IP address.
The use-case is when using simple-vpn within another private subnet, using IPv6-only makes it easy to pick an IP block that won't clash with the rest of the subnet.
Would you be able to support this use-case? Open to PRs?

I was trying to setup the project locally and I get the following error when I try to get dependencies:
```
$ go get ./...
# simple-vpn
./cmd_server.go:285:11: tapConfig.Name undefined (type water.Config has no field or method Name)
```
Did the `water` library change their interface for tapConfig? It might be worth using `dep` to lock down the dependencies to specific versions

Ahh Mac OS X is probably going to cause you some pain :/

I see the same thing:

      frodo ~/go/src/github.com/skx/simple-vpn $ GOOS=darwin GOARCH=amd64 go build .
      ./cmd_server.go:285:11: tapConfig.Name undefined (type water.Config has no field or method Name)

Looks like that's noted in the source:

* https://github.com/songgao/water/blob/master/syscalls_darwin.go#L127

--- a/.travis/build-all.sh
+++ b/.travis/build-all.sh
@@ -6,43 +6,32 @@ BASE="simple-vpn"
 # Get the dependencies
 go get -t -v -d $(go list ./...)
 
+
 #
-# We could build on multiple platforms/archs.
+# Function to do a build
 #
-# Except we use CGO, so we can't.
+function do_build {
+
+    export GOOS=$1
+    export GOARCH=$2
+
+    if [ "${GOARCH}" = "arm64" ]; then
+        export GOOS=""
+        export GOARM=7
+    fi
+
+    OUT=$3
+    echo "OUT: $OUT"
+    go build -ldflags "-X main.version=$(git describe --tags)" -o "${OUT}"
+}
+
 #
-BUILD_PLATFORMS="linux"
-BUILD_ARCHS="amd64"
+# Linux
+#
+do_build linux amd64 "${BASE}-linux-amd64"
+do_build linux 386 "${BASE}-linux-i386"
 
-# For each platform
-for OS in ${BUILD_PLATFORMS[@]}; do
-
-    # For each arch
-    for ARCH in ${BUILD_ARCHS[@]}; do
-
-        # Setup a suffix for the binary
-        SUFFIX="${OS}"
-
-        # i386 is better than 386
-        if [ "$ARCH" = "386" ]; then
-            SUFFIX="${SUFFIX}-i386"
-        else
-            SUFFIX="${SUFFIX}-${ARCH}"
-        fi
-
-        # Windows binaries should end in .EXE
-        if [ "$OS" = "windows" ]; then
-            SUFFIX="${SUFFIX}.exe"
-        fi
-
-        echo "Building for ${OS} [${ARCH}] -> ${BASE}-${SUFFIX}"
-
-        # Run the build
-        export GOARCH=${ARCH}
-        export GOOS=${OS}
-        export CGO_ENABLED=1
-
-        go build -ldflags "-X main.version=$(git describe --tags)" -o "${BASE}-${SUFFIX}"
-
-    done
-done
+#
+# ARM
+#
+do_build arm64 arm64  "${BASE}-arm64"

--- a/cmd_client.go
+++ b/cmd_client.go
@@ -55,12 +55,22 @@ func (p *clientCmd) configureClient(dev *water.Interface, ip string, subnet stri
 	devStr := dev.Name()
 
 	//
+	// Ensure we have the right mask for the client IP
+	//
+	fmt.Printf("Client IP is %s\n", ip)
+	if strings.Contains(ip, ":") {
+		ip += "/32"
+	} else {
+		ip += "/128"
+	}
+
+	//
 	// The commands we're going to execute
 	//
 	cmds := [][]string{
 		{"ip", "link", "set", "dev", devStr, "up"},
 		{"ip", "link", "set", "mtu", mtuStr, "dev", devStr},
-		{"ip", "addr", "add", ip + "/32", "dev", devStr},
+		{"ip", "addr", "add", ip, "dev", devStr},
 		{"ip", "route", "add", gateway, "dev", devStr},
 		{"ip", "route", "add", subnet, "via", gateway},
 	}

--- a/cmd_client.go
+++ b/cmd_client.go
@@ -59,9 +59,9 @@ func (p *clientCmd) configureClient(dev *water.Interface, ip string, subnet stri
 	//
 	fmt.Printf("Client IP is %s\n", ip)
 	if strings.Contains(ip, ":") {
-		ip += "/32"
-	} else {
 		ip += "/128"
+	} else {
+		ip += "/32"
 	}
 
 	//

--- a/cmd_client.go
+++ b/cmd_client.go
@@ -370,7 +370,7 @@ func (p *clientCmd) Execute(_ context.Context, f *flag.FlagSet, _ ...interface{}
 		return nil
 	})
 
-	socket.Serve()
+	socket.Serve(false)
 	socket.Wait()
 
 	return subcommands.ExitSuccess

--- a/cmd_server.go
+++ b/cmd_server.go
@@ -558,6 +558,10 @@ func (p *serverCmd) serveWs(w http.ResponseWriter, r *http.Request) {
 	//    1.2.3.0       -> (internal) IP of VPN-server
 	//
 	socket.SendCommand("init", p.subnet, clientIP, fmt.Sprintf("%d", p.mtu), p.serverIP)
-	socket.Serve()
+
+	//
+	// IPv6 requires different handling.  Sigh.
+	//
+	socket.Serve(strings.Contains(p.serverIP, ":"))
 	socket.Wait()
 }

--- a/cmd_server.go
+++ b/cmd_server.go
@@ -167,7 +167,8 @@ func (p *serverCmd) pickIP(name string, remote string) (string, error) {
 		s := ip.String()
 
 		// Skip the first IP.
-		if strings.HasSuffix(s, ".0") {
+		if strings.HasSuffix(s, ".0") ||
+			strings.HasSuffix(s, ":") {
 			continue
 		}
 
@@ -242,30 +243,41 @@ func (p *serverCmd) Execute(_ context.Context, f *flag.FlagSet, _ ...interface{}
 	}
 
 	//
+	// Here we used to mark every IP in the network range
+	// as being allocated.
+	//
+	// Instead we'll only claim the first, which will be the
+	// server IP.
+	//
 	// For each IP in the range we now mark the IP as free.
 	//
 	p.assigned = make(map[string]*connection)
-	for ip := ip.Mask(subnet.Mask); subnet.Contains(ip); incIP(ip) {
+	for ip := ip.Mask(subnet.Mask); subnet.Contains(ip) && p.serverIP == ""; incIP(ip) {
 
 		s := ip.String()
 
-		if strings.HasSuffix(s, ".0") {
+		// Skip anything that ends in `.0`, or `:`.
+		if strings.HasSuffix(s, ".0") ||
+			strings.HasSuffix(s, ":") {
 			continue
 		}
 
-		p.assigned[s] = nil
-
 		//
-		// The first IP in the range is the server's IP.
+		// OK we've got the IP for the server
 		//
-		if p.serverIP == "" {
-			p.serverIP = s
+		p.serverIP = s
+		p.assigned[s] = &connection{localIP: s, remoteIP: s, name: "vpn-server"}
+		fmt.Printf("VPN server has IP %s\n", p.serverIP)
 
-			fmt.Printf("VPN server has IP %s\n", p.serverIP)
+	}
 
-			// Mark this IP as being unavailable
-			p.assigned[s] = &connection{localIP: s, remoteIP: s, name: "vpn-server"}
-		}
+	//
+	// Are we using IPv6?
+	//
+	if strings.Contains(p.serverIP, ":") {
+		fmt.Printf("VPN server using IPv6.\n")
+	} else {
+		fmt.Printf("VPN server using IPv4.\n")
 	}
 
 	//

--- a/etc/client.cfg
+++ b/etc/client.cfg
@@ -47,7 +47,7 @@ key = Iequa[oogho5reiNgoo7ci4ruho~r#%fdsflj30-1l;alj1.>SDF£LK!
 ## variables:
 ##
 ##  $DEVICE    -  e.g. tun0.
-##  $CLIENT_IP -  e.g 10.0.0.2.
+##  $CLIENT_IP -  e.g. 10.0.0.2.
 ##  $SERVER_IP -  e.g. 10.0.0.1.
 ##  $SUBNET    -  e.g. 10.0.0.0/24.
 ##  $MTU       -  e.g. 1280.
@@ -76,6 +76,8 @@ key = Iequa[oogho5reiNgoo7ci4ruho~r#%fdsflj30-1l;alj1.>SDF£LK!
 ##     {"Name":"alert.vpn","IP":"10.137.248.3"} ]
 ##
 ##
+#
+# Here we just dump them to the console.
 #
 # peers = /bin/cat
 #

--- a/etc/server.cfg
+++ b/etc/server.cfg
@@ -35,6 +35,10 @@ key = Iequa[oogho5reiNgoo7ci4ruho~r#%fdsflj30-1l;alj1.>SDF£LK!
 ## common.  But IPv6 is the way of the future, so if you wish you can
 ## allow your connected clients to communicate over IPv6.
 ##
+## NOTE: That our IPv6 implementation is naive, and ignores
+##       router-advertisement, neighbour solications, etc.
+##       IPv6 traffic is essentially broadcast to all peers.
+##
 ## You'll need to pick a private IPv6 range, somewhere beneath the
 ## range fde4:8dba:82e1::/48, for example here we use the hungry-food
 ## range:

--- a/etc/server.cfg
+++ b/etc/server.cfg
@@ -25,11 +25,22 @@ key = Iequa[oogho5reiNgoo7ci4ruho~r#%fdsflj30-1l;alj1.>SDF£LK!
 
 
 ##
-## Change the subnet from which we allocate IPs
+## Change the subnet from which we allocate client IP addresses.
 ##
 #
 # subnet = 10.137.248.0/24
 #
+##
+## IPv4 is more efficient (as the header size is smaller), and more
+## common.  But IPv6 is the way of the future, so if you wish you can
+## allow your connected clients to communicate over IPv6.
+##
+## You'll need to pick a private IPv6 range, somewhere beneath the
+## range fde4:8dba:82e1::/48, for example here we use the hungry-food
+## range:
+##
+## subnet = fde4:8dba:82e1:food::/64
+##
 
 
 ##

--- a/etc/server.cfg
+++ b/etc/server.cfg
@@ -39,7 +39,7 @@ key = Iequa[oogho5reiNgoo7ci4ruho~r#%fdsflj30-1l;alj1.>SDF£LK!
 ## range fde4:8dba:82e1::/48, for example here we use the hungry-food
 ## range:
 ##
-## subnet = fde4:8dba:82e1:food::/64
+## subnet = fde4:8dba:82e1:f00d::/64
 ##
 
 
@@ -52,8 +52,10 @@ key = Iequa[oogho5reiNgoo7ci4ruho~r#%fdsflj30-1l;alj1.>SDF£LK!
 ##
 #
 # host_deagol = 10.137.248.10
+# host_deagol = fde4:8dba:82e1:f00d::10
 #
 # host_frodo  = 10.137.248.20
+# host_deagol = fde4:8dba:82e1:f00d::20
 #
 
 

--- a/shared/socketHandler.go
+++ b/shared/socketHandler.go
@@ -1,3 +1,17 @@
+// shared/socketHandler.go contains code to route traffic over
+// the socket.
+//
+// When a client connects to the VPN server their local tun device
+// is one end of the socket, and the WS-connection is the other.
+//
+// For the server we have an array of such things, and we handle
+// traffic by sending to the "correct" socket by MAC address - except
+// in the case of IPv6 where we broadcast.
+//
+// IPv6 behaviour could, and should, be improved.  But handling router
+// advertisements, neighbour solicitations, etc, is hard.  Better to
+// keep it simple.  Keep it secret.  Keep it safe.
+
 package shared
 
 import (
@@ -263,6 +277,10 @@ func (s *Socket) Serve(ipv6 bool) {
 
 				if len(msg) >= 14 {
 
+					//
+					// IPv4 ttaffic involves routing
+					// correctly.
+					//
 					if ipv6 == false {
 						s.setMACFrom(msg)
 
@@ -281,6 +299,10 @@ func (s *Socket) Serve(ipv6 bool) {
 						}
 					} else {
 
+						//
+						// IPv6 traffic is just
+						// broadcast as-is.
+						//
 						BroadcastMessage(websocket.BinaryMessage, msg, s)
 					}
 				}
@@ -290,7 +312,7 @@ func (s *Socket) Serve(ipv6 bool) {
 				}
 				s.iface.Write(msg)
 			} else if msgType == websocket.TextMessage {
-				// in-band stuff.
+				// in-band messages over the WS link
 
 				str := strings.Split(string(msg), "|")
 				if len(str) < 2 {

--- a/shared/socketHandler.go
+++ b/shared/socketHandler.go
@@ -236,7 +236,7 @@ func (s *Socket) tryServeIfaceRead() {
 
 // Serve is the main-driver which never returns
 // Handle proxying data back and forth..
-func (s *Socket) Serve() {
+func (s *Socket) Serve(ipv6 bool) {
 	s.writeLock.Lock()
 	defer s.writeLock.Unlock()
 	s.tryServeIfaceRead()
@@ -262,19 +262,25 @@ func (s *Socket) Serve() {
 			if msgType == websocket.BinaryMessage {
 
 				if len(msg) >= 14 {
-					s.setMACFrom(msg)
 
-					dest := GetDestMAC(msg)
-					isUnicast := MACIsUnicast(dest)
+					if ipv6 == false {
+						s.setMACFrom(msg)
 
-					var sd *Socket
-					if isUnicast {
-						sd = FindSocketByMAC(dest)
-						if sd != nil {
-							sd.WriteMessage(websocket.BinaryMessage, msg)
-							continue
+						dest := GetDestMAC(msg)
+						isUnicast := MACIsUnicast(dest)
+
+						var sd *Socket
+						if isUnicast {
+							sd = FindSocketByMAC(dest)
+							if sd != nil {
+								sd.WriteMessage(websocket.BinaryMessage, msg)
+								continue
+							}
+						} else {
+							BroadcastMessage(websocket.BinaryMessage, msg, s)
 						}
 					} else {
+
 						BroadcastMessage(websocket.BinaryMessage, msg, s)
 					}
 				}

--- a/shared/util.go
+++ b/shared/util.go
@@ -3,17 +3,17 @@ package shared
 // MacAddr stores a MAC address.
 type MacAddr [6]byte
 
-// GetSrcMAC retreives the source MAC address of a TCP/IP packet.
+// GetSrcMAC retrieves the source MAC address of a TCP/IP packet.
 func GetSrcMAC(packet []byte) MacAddr {
 	var mac MacAddr
-	copy(mac[:], packet[6:12])
+	copy(mac[:], packet[16:32])
 	return mac
 }
 
 // GetDestMAC retreives the destination MAC address of a TCP/IP packet.
 func GetDestMAC(packet []byte) MacAddr {
 	var mac MacAddr
-	copy(mac[:], packet[0:6])
+	copy(mac[:], packet[0:16])
 	return mac
 }
 

--- a/shared/util.go
+++ b/shared/util.go
@@ -6,14 +6,28 @@ type MacAddr [6]byte
 // GetSrcMAC retrieves the source MAC address of a TCP/IP packet.
 func GetSrcMAC(packet []byte) MacAddr {
 	var mac MacAddr
-	copy(mac[:], packet[16:32])
+
+	if 6 == (packet[0] >> 4) {
+		// ipv6
+		copy(mac[:], packet[16:32])
+	} else {
+		// ipv4
+		copy(mac[:], packet[6:12])
+	}
 	return mac
 }
 
 // GetDestMAC retreives the destination MAC address of a TCP/IP packet.
 func GetDestMAC(packet []byte) MacAddr {
 	var mac MacAddr
-	copy(mac[:], packet[0:16])
+
+	if 6 == (packet[0] >> 4) {
+		// ipv6
+		copy(mac[:], packet[0:16])
+	} else {
+		// ipv4
+		copy(mac[:], packet[0:6])
+	}
 	return mac
 }
 

--- a/shared/util.go
+++ b/shared/util.go
@@ -6,28 +6,14 @@ type MacAddr [6]byte
 // GetSrcMAC retrieves the source MAC address of a TCP/IP packet.
 func GetSrcMAC(packet []byte) MacAddr {
 	var mac MacAddr
-
-	if 6 == (packet[0] >> 4) {
-		// ipv6
-		copy(mac[:], packet[16:32])
-	} else {
-		// ipv4
-		copy(mac[:], packet[6:12])
-	}
+	copy(mac[:], packet[6:12])
 	return mac
 }
 
 // GetDestMAC retreives the destination MAC address of a TCP/IP packet.
 func GetDestMAC(packet []byte) MacAddr {
 	var mac MacAddr
-
-	if 6 == (packet[0] >> 4) {
-		// ipv6
-		copy(mac[:], packet[0:16])
-	} else {
-		// ipv4
-		copy(mac[:], packet[0:6])
-	}
+	copy(mac[:], packet[0:6])
 	return mac
 }
 

